rules/move_on_aptos/objects/security/missing-ownership-check.yaml

```diff
@@ -17,58 +18,220 @@ rules:
+                    aptos_framework::fungible_asset::Metadata
+        by-side-effect: true
     pattern-sanitizers:
+        #TODO
+        #1. use regex
+        #2. use propagator
       - patterns:
-          - pattern-either:
-            # standalone checker
-            # TODO: if interfile analysis is not enabled, consider adding some safe
-            #       library functions to the standalone checker list, such as:
-            - pattern: aptos_framework::code::freeze_code_object(..., $OBJ, ...)
-            - pattern: std::code::freeze_code_object(..., $OBJ, ...)
-            - pattern: fungible_asset::deposit_sanity_check($OBJ, ...)
-
-            # conditional checker
-            - patterns:
-              # scope: inside assert() or if() block
-              - pattern-either:
-                  - pattern-inside: assert!(...)
-                  - pattern-inside: if (...) { ... }
-              # checker:
-              - pattern-either:
-                  # specific patterns
-                  - pattern: aptos_framework::object::owner(& $OBJ)
-                  - pattern: std::object::owner(& $OBJ)
-
-                  - pattern: aptos_framework::object::is_owner($OBJ, ...)
-                  - pattern: std::object::is_owner($OBJ, ...)
-
-                  - pattern: aptos_framework::object::owns($OBJ, ...)
-                  - pattern: std::object::owns($OBJ, ...)
-
-                  - pattern: collection::creator($OBJ)
-          - focus-metavariable: $OBJ
+        - pattern-either:
+            # aptos_framework::object
+          - pattern: |
+              assert!(aptos_framework::object::owner<$TYPE>($OBJ) == aptos_framework::signer::address_of($SIGNER),$ERROR_CODE);
+          - pattern: |
+              let $ADDRESS_OF_SIGNER = aptos_framework::signer::address_of($SIGNER);
+              ...
+              assert!(aptos_framework::object::owner<$TYPE>($OBJ) == $ADDRESS_OF_SIGNER,$ERROR_CODE);
+          - pattern: |
+              let $ADDRESS_OF_SIGNER = aptos_framework::signer::address_of($SIGNER);
+              ...
+              let $OBJECT_OWNER = aptos_framework::object::owner<$TYPE>($OBJ);
+              ...
+              assert!($OBJECT_OWNER == $ADDRESS_OF_SIGNER,$ERROR_CODE);
+          - pattern: |
+              let $OBJECT_OWNER = aptos_framework::object::owner<$TYPE>($OBJ);
+              ...
+              let $ADDRESS_OF_SIGNER = aptos_framework::signer::address_of($SIGNER);
+              ...
+              assert!($OBJECT_OWNER == $ADDRESS_OF_SIGNER,$ERROR_CODE);
+          - pattern: |
+              aptos_framework::object::transfer<$TYPE>($SIGNER, $OBJ, $ANY_VAR);
+          - pattern: |
+              aptos_framework::object::transfer_to_object<$TYPE>($SIGNER,$OBJ, $ANY_VAR);
+
+            #start test
+          - pattern-inside: |
+              aptos_framework::object::is_owner<$TYPE>($OBJ, aptos_framework::signer::address_of($SIGNER))
+        
+          - pattern-inside: |
```

Comment:
Why do we have some `pattern` and some `pattern-inside` matching a similar pattern? Can't we just use `pattern`?

Reply:
Matching `aptos_framework::object::is_owner` is actually broken. I was trying different patterns, but no one worked as expected